central/image/service/service_impl.go

```diff
@@ -568,11 +568,12 @@ func (s *serviceImpl) EnrichLocalImageInternal(ctx context.Context, request *v1.
 					// to help troubleshoot potential image name or caching issues log the request's image
 					// name as well.
 					logging.String("request_image", request.GetImageName().GetFullName()),
+					logging.String("request_id", request.GetRequestId()),
 				)
 
-				if imgExists {
-					// In case we hit an error during enriching, and the image previously existed, we will _not_ upsert it in
-					// central, since it could lead to us overriding an enriched image with a non-enriched image.
+				if imgExists || request.GetRequestId() != "" {
```

Comment:
Help me understand this: It's possible that the metadata was retrieved successfully, but since this is "ad-hoc" (ie. API call driven scan), we are OK?

Also, wondering if we need to check for API-call drive scans for signatures as well (line 587).

Reply:
> Help me understand this: It's possible that the metadata was retrieved successfully, but since this is "ad-hoc" (ie. API call driven scan), we are OK?

Right, for example: user executes `roxctl image scan ...`, the request is delegated to the secured cluster, metadata retrieved successfully, image indexed successfully, matching request sent to Central but fails due to vuln data not being loaded yet. Current state a 'successful' response is returned to the API/roxctl, but when the JSON is inspected the scan is empty. This is wrong and an error should be returned - otherwise CI or other systems will believe there are no vulnerabilities detected and the image is clean.

> Also, wondering if we need to check for API-call drive scans for signatures as well (line 587).

Good callout - my first reaction is to leave it be until there is a reason to change it + given how close to release this is and being unclear how to test the failure scenario.

Currently in Sensor no [error is returned when sig pulling fails](https://github.com/stackrox/stackrox/blob/4575c4a00c790f13c41226c952968ab2357d245a/sensor/common/scan/scan.go#L378-L401), if 'continuing' that behavior we shouldn't error out in Central either when there are sig validation errors. That being said the chance of an error occurring for sig validation is much lower since there are no RPC calls being made to validate signatures.  On the flip side Central enrichments will fail if there is a sig validation error - so the two are not in sync currently.  We did have an effort starting to revisit the enrichers because there are many nuances where the behaviors are not the same.

I believe the way delegated scanning is currently working makes 'more' sense because a `roxctl image scan` probably shouldn't error out when sig validation fails (is no reason to), however a `roxctl image check` should fail  if sig validation fails - but only when there is policy checking sigs.

TL;DR; is some debt here already - will add this to the list of items to revisit when improving the enrichers, but for right now recommend leaving it as.